src/emplacer.rs

```diff
@@ -0,0 +1,288 @@
+        let parcel_path = self.cache_path.join(&parcel.label.sha256);
+        if parcel_path.is_file() {
+            return Ok(());
+        }
+
+        let parcel_data = reader.get_parcel(invoice_id, &parcel).await?;
+        safely_write(&parcel_path, parcel_data).await
+            .with_context(|| format!("Error caching parcel {} at {}", parcel.label.name, parcel_path.display()))
+    }
+
+    async fn emplace_as_asset(&self, reader: &impl BindleReader, invoice_id: &bindle::Id, parcel: &bindle::Parcel) -> anyhow::Result<()> {
+        let parcel_path = self.asset_parcel_path(invoice_id, parcel);
+        if parcel_path.is_file() {
+            return Ok(());
+        }
+
+        let parcel_data = reader.get_parcel(invoice_id, &parcel).await?;
+        safely_write(&parcel_path, parcel_data).await
+            .with_context(|| format!("Error caching parcel {} at {}", parcel.label.name, parcel_path.display()))?;
+        Ok(())
+    }
+
+    async fn emplace_as_assets(&self, reader: &impl BindleReader, invoice_id: &bindle::Id, parcels: &Vec<bindle::Parcel>) -> anyhow::Result<()> {
+        let placement_futures = parcels.iter().map(|parcel| self.emplace_as_asset(reader, invoice_id, parcel));
+        let all_placements = futures::future::join_all(placement_futures).await;
+        let first_error = all_placements.into_iter().find(|p| p.is_err());
+        first_error.unwrap_or(Ok(()))
+    }
+
+    // TODO: there is a potential risk here if two bindle servers have different content
+    // for the same invoice id - if we cached data from the 'old' server we would use that
+    // in place of the new one
+    fn invoice_path(&self, invoice_id: &bindle::Id) -> PathBuf {
+        let filename = invoice_cache_key(invoice_id);
+        self.invoices_path().join(filename)
+    }
+
+    fn module_parcel_path(&self, parcel: &bindle::Parcel) -> PathBuf {
+        self.cache_path.join(&parcel.label.sha256)
+    }
+
+    fn asset_parcel_path(&self, invoice_id: &bindle::Id, parcel: &bindle::Parcel) -> PathBuf {
+        self.asset_path_for(invoice_id).join(&parcel.label.name)
+    }
+
+    fn invoices_path(&self) -> PathBuf {
+        self.cache_path.join("_INVOICES")
+    }
+
+    fn asset_path(&self) -> PathBuf {
+        self.cache_path.join("_ASSETS")
+    }
+
+    pub fn asset_path_for(&self, invoice_id: &bindle::Id) -> PathBuf {
+        let key = invoice_cache_key(invoice_id);
+        self.asset_path().join(key)
+    }
+
+    fn asset_dir_volume_mount(&self, invoice_id: &bindle::Id) -> HashMap<String, String> {
+        let mut volumes = HashMap::new();
+        volumes.insert("/".to_owned(), self.asset_path_for(invoice_id).display().to_string());  // TODO: maybe volumes should map PathBufs // or struct of host and guest
+        volumes
+    }
+    
+}
+
+pub(crate) fn invoice_cache_key(id: &bindle::Id) -> String {
```

Comment:
Same re: being part of a `bindle_utils` module.

Reply:
This is specific to the way the emplacer organises the file cache.  It's not a generic bindle helper.  However, it should be private to the emplacer module - it was temporarily public because it was temporarily shared with the old code.  Thanks!